File: LeetCode/Medium/0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.js
/*
P:
    input: arr of strings (class commands), 2-D array (matrix of actions)
    output: arr of booleans + integers (responses of methods)
    task: build out the methods for this RandomizedSet class:
            - insert: inserts a integer in the set, returns boolean if it was successful or not
            - delete: removes a integer in the set, returns boolean if it was successful or not
            - getRandom: returns a random element from the current set we have
        and each method must have a constant time complexity [ O(1) ]
E: 

D:
    arrays, integers, classes, objects, set (?)

A:

C:
*/

class RandomizedSet {
    constructor(set = new Set()) {
        this.set = set;
    }

    insert(val) {
        if (this.set.has(val)) {
            return false;
        } else {
            this.set.add(val);
            return true;
        }
    }

    remove(val) {
        if (!this.set.has(val)) {
            return false;
        } else {
            this.set.delete(val);
            return true;
        }
    }
    
    getRandom() {
        const setArr = [...this.set];

        const randomSetIndex = Math.floor(Math.random() * setArr.length);

        return setArr[randomSetIndex];


    }
}

// var RandomizedSet = function() {
    
// };

// /** 
//  * @param {number} val
//  * @return {boolean}
//  */
// RandomizedSet.prototype.insert = function(val) {
    
// };

// /** 
//  * @param {number} val
//  * @return {boolean}
//  */
// RandomizedSet.prototype.remove = function(val) {
    
// };

// /**
//  * @return {number}
//  */
// RandomizedSet.prototype.getRandom = function() {
    
// };

/** 
 * Your RandomizedSet object will be instantiated and called as such:
 * var obj = new RandomizedSet()
 * var param_1 = obj.insert(val)
 * var param_2 = obj.remove(val)
 * var param_3 = obj.getRandom()
 */

File: LeetCode/Medium/0380-insert-delete-getrandom-o1/0380-insert-delete-getrandom-o1.py
import random, math

class RandomizedSet:

    def __init__(self):
        self.arr = []
        self.hash_map = {}
        
    def insert(self, val: int) -> bool:
        if val in self.hash_map:
            return False
        else:
            self.arr.append(val)
            self.hash_map[val] = len(self.arr) - 1
            return True


    def remove(self, val: int) -> bool:
        if val not in self.hash_map:
            return False
        else:
            recent_val = self.arr[-1]
            val_index = self.hash_map[val]
            recent_val_index = self.hash_map[recent_val]

            # index swap in array
            self.arr[-1] = val
            self.arr[val_index] = recent_val

            # index swap in hash map
            self.hash_map[recent_val] = val_index
            self.hash_map[val] = recent_val_index

            # removal of value (O(1) time complexity)
            self.arr.pop()
            del self.hash_map[val]

            return True
        
    def getRandom(self) -> int:
        random_index = random.randint(0,  len(self.arr) - 1)

        return self.arr[random_index]
        


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()